src/apps/dictionaries/pages/AddBulkConceptsPage.tsx

```diff
@@ -22,22 +40,147 @@ const useStyles = makeStyles((theme: Theme) =>
       textAlign: "center",
       marginTop: "2vh"
     },
-    lightColour: {
-      color: theme.palette.background.default
+    switchButton: {
+      color: "white !important"
+    },
+    sourcesDropdownHeader: {
+      padding: "0.5rem 1rem"
+    },
+    input: {
+      cursor: "pointer",
+      borderBottom: "1px dotted black",
+      paddingBottom: "0.25rem"
+    },
+    underline: {
+      "&&&:before": {
+        borderBottom: "none"
+      },
+      "&&:after": {
+        borderBottom: "none"
+      }
+    },
+    sourceIcon: {
+      marginRight: "0.2rem",
+      fill: "#8080809c"
+    },
+    searchInput: {
+      textAlign: "center",
+      fontSize: "larger"
+    },
+    csvLink: {
+      marginTop: "0.5rem"
+    },
+    newPageIcon: {
+      color: "white",
+      marginTop: "0.5rem",
+      width: "1rem",
+      height: "1rem"
+    },
+    loadMoreButton: {
+      margin: "auto",
+      display: "flex",
+      border: "none",
+      fontWeight: "bold",
+      marginTop: "1rem",
+      marginBottom: "1rem"
+    },
+    seenAllText: {
+      margin: "auto",
+      fontWeight: "bold",
+      marginTop: "1rem",
+      marginBottom: "1rem",
+      textAlign: "center"
     }
   })
 );
 
 interface Props {
+  sources: APISource[];
+  meta?: { num_found?: number };
   addConceptsToDictionary: (
     ...args: Parameters<typeof recursivelyAddConceptsToDictionaryAction>
   ) => void;
+  retrievePublicSources: (
+    ...args: Parameters<typeof retrievePublicSourcesAction>
+  ) => Promise<any>;
 }
 
-const AddBulkConceptsPage: React.FC<Props> = ({ addConceptsToDictionary }) => {
+const AddBulkConceptsPage: React.FC<Props> = ({ addConceptsToDictionary, sources = [], retrievePublicSources, meta = {} }) => {
   const classes = useStyles();
   const { pathname: url } = useLocation();
   const { fromSource } = useQueryParams();
+  const history = useHistory();
+
+  
+  const [showAllSources, setShowAllSources] = useState(false);
+
+  const [queryString, setQueryString] = useState('');
+  const handleSearch = (q: string) => setQueryString(q);
+
+  const [currentSources, setCurrentSources] = useState<
+    { name: string; sourceUrl: string }[]
+  >([]);
+
+  const defaultSources = Object.entries(
+    PREFERRED_SOURCES
+  ).map(([key, value]) => ({ name: key, sourceUrl: value }));
+
+  const allSources = defaultSources.concat(sources.map(s => ({ name: s.name, sourceUrl: s.url })))
```

Comment:
Since we're re-writing this anonymous function all over the place here, it seems like we should turn it into a named function:

```ts
  s => ({ name: s.name, sourceUrl: s.url })
```

```ts
const mapSources = (s: APISource) => ({
  name: s.name,
  sourceUrl: s.url
})
```

Then replace the other calls with just `mapSources`.